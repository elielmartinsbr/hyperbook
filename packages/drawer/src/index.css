#drawer-root {
  font-family: hyperbook-body, sans-serif;
}

#drawer-root * {
  box-sizing: border-box;
}

.drawer {
  background: #fff;
  font-size: 16px;
}

.drawer-container {
  --transition-speed: 0.3s;
}

.drawer {
  width: 300px;
  height: 100%;
  overflow: auto;
  position: fixed;
  display: flex;
  box-shadow: 0 0 15px rgba(0, 0, 0, 0.5);
  transition: transform var(--transition-speed) ease;
  z-index: 1000;
}

.drawer.left {
  top: 80px;
  left: 0;
  transform: translateX(-105%);
}

.drawer.right {
  top: 80px;
  right: 0;
  transform: translateX(100%);
}

.drawer-container.in.open .left,
.drawer-container.in.open .right {
  transform: translateX(0);
}

.backdrop {
  visibility: hidden;
  opacity: 0;
  background: rgba(0, 0, 0, 0.5);
  transition: opacity var(--transition-speed) ease,
    visibility var(--transition-speed) ease;
  width: 100%;
  height: 100%;
  top: 80px;
  left: 0;
  position: fixed;
  pointer-events: none;
  z-index: 0;
}

.drawer-container.in.open .backdrop {
  visibility: visible;
  opacity: 1;
  pointer-events: auto;
  z-index: 999;
}
