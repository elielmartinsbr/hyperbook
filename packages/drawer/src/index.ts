import "./index.css";
export * from "./Drawer";
