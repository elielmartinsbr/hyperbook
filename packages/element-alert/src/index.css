.hyperbook.element-alert > p {
  margin-bottom: 0;
}

.hyperbook.element-alert {
  letter-spacing: 0;
  box-sizing: border-box;
  font-size: inherit;
  line-height: 1.6rem;
  word-spacing: 0.05rem;
  border-bottom-right-radius: 2px;
  border-top-right-radius: 2px;
  padding: 8px 12px 8px 20px;
  margin-bottom: 16px;
  position: relative;
  border-left-width: 4px;
  border-left-style: solid;
}

.hyperbook.element-alert::before {
  border-radius: 100%;
  font-size: 14px;
  font-weight: 700;
  left: -12px;
  line-height: 20px;
  display: flex;
  align-items: center;
  justify-content: center;
  position: absolute;
  height: 20px;
  width: 20px;
  text-align: center;
  top: 12px;
}

.hyperbook.element-alert.success {
  border-left-style: solid;
  border-left-width: 4px;
}

.hyperbook.element-alert.success::before {
  content: "âœ“";
}

.hyperbook.element-alert.info {
  border-left-style: solid;
  border-left-width: 4px;
}

.hyperbook.element-alert.info::before {
  content: "i";
}

.hyperbook.element-alert.warn {
  border-left-style: solid;
  border-left-width: 4px;
}

.hyperbook.element-alert.warn::before {
  content: "!";
}

.hyperbook.element-alert.error {
  border-left-style: solid;
  border-left-width: 4px;
}

.hyperbook.element-alert.error::before {
  content: "x";
}

.hyperbook.element-alert {
  background-color: var(--color-nav);
  border-left-color: var(--color-nav-border);
}

.hyperbook.element-alert::before {
  color: #fff;
}

.hyperbook.element-alert.success {
  border-left-color: #00c851;
}

.hyperbook.element-alert.success::before {
  background-color: #00c851;
}

.hyperbook.element-alert.info {
  border-left-color: #33b5e5;
}

.hyperbook.element-alert.info::before {
  background-color: #33b5e5;
}

.hyperbook.element-alert.warn {
  border-left-color: #f0b429;
}

.hyperbook.element-alert.warn::before {
  background-color: #f0b429;
}

.hyperbook.element-alert.error {
  border-left-color: #ef4e4e;
}

.hyperbook.element-alert.error::before {
  background-color: #ef4e4e;
}
