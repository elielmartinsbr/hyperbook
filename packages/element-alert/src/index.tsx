import { FC, ReactNode } from "react";
import "./index.css";

type DirectiveAlertProps = {
  children?: ReactNode;
  node: any;
} & any;

const DirectiveAlert: FC<DirectiveAlertProps> = ({
  children,
  node,
  ...props
}) => {
  let c = Object.keys(props).join(" ");
  return <div className={`hyperbook element-alert ${c}`}>{children}</div>;
};

export default {
  directives: { alert: DirectiveAlert },
};
