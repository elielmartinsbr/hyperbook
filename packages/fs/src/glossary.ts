import { Glossary } from "@hyperbook/types";
import { getMarkdown, listForFolder, VFile } from "./vfile";

export const get = async (root: string): Promise<Glossary> => {
  const glossary: Glossary = {};
  const files = await listForFolder(root, "glossary");
  for (const file of files) {
    const { data } = await getMarkdown(file);
    let name = data.name;
    if (file.path.href === null) {
      continue;
    }

    if (!name) {
      name = file.name;
    }

    const letter = name[0].toUpperCase();
    if (!glossary[letter]) {
      glossary[letter] = [];
    }

    glossary[letter].push({
      name,
      href: file.path.href,
    });
  }
  return glossary;
};

export const getReferences = async (file: VFile): Promise<VFile[]> => {
  const bookFiles = await listForFolder(file.root, "book");
  const vfiles: VFile[] = [];
  for (const bookFile of bookFiles) {
    const { content, data } = await getMarkdown(bookFile);
    const r = new RegExp(
      `:t\\[.*\\]\\{#${file.name}(\..*)?\\}|:t\\[${file.name}\\]`
    );
    const m = content.match(r);
    if (m && !data.hide) {
      vfiles.push({ ...bookFile, name: data.name || bookFile.name });
    }
  }
  return vfiles;
};
