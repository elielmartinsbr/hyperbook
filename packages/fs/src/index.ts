import * as hyperproject from "./hyperproject";
import * as hyperlibrary from "./hyperlibrary";
import * as hyperbook from "./hyperbook";
import * as glossary from "./glossary";
import * as vfile from "./vfile";

export { VFile } from "./vfile";

export { hyperlibrary, hyperbook, glossary, hyperproject, vfile };
