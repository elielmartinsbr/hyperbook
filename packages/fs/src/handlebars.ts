//@ts-ignore
import handlebars from "handlebars/dist/cjs/handlebars";
import { findUpSync, Options } from "find-up";
import { lookup } from "mime-types";
import fs from "fs";
import path from "path";
import { extractLines } from "./vfile";

const isString = (s: any): s is string => {
  return typeof s === "string";
};

const chop = (str: string): string => {
  if (!isString(str)) return "";
  var re = /^[-_.\W\s]+|[-_.\W\s]+$/g;
  return str.trim().replace(re, "");
};

const changecase = (str: string, fn: (s: string) => string): string => {
  if (!isString(str)) return "";
  if (str.length === 1) {
    return str.toLowerCase();
  }

  str = chop(str).toLowerCase();

  var re = /[-_.\W\s]+(\w|$)/g;
  return str.replace(re, function (_, ch) {
    return fn(ch);
  });
};

handlebars.registerHelper("times", (n: number, block: any) => {
  let accum = "";
  for (let i = 0; i < n; ++i) accum += block.fn(i);
  return accum;
});

handlebars.registerHelper("concat", (s1: string, s2: string) => {
  return s1 + s2;
});

handlebars.registerHelper("camelcase", (s: string) => {
  if (!isString(s)) return "";
  return changecase(s, (c) => c.toUpperCase());
});

handlebars.registerHelper("dashcase", (s: string) => {
  if (!isString(s)) return "";
  return changecase(s, (c) => "-" + c);
});

handlebars.registerHelper("lowercase", (s: string) => {
  if (!isString(s)) return "";
  return s.toLowerCase();
});

handlebars.registerHelper("uppercase", (s: string) => {
  if (!isString(s)) return "";
  return s.toUpperCase();
});

handlebars.registerHelper("pascalcase", (s: string) => {
  if (!isString(s)) return "";
  s = changecase(s, (c) => c.toUpperCase());
  return s.charAt(0).toUpperCase() + s.slice(1);
});

handlebars.registerHelper("replaceAll", (s: string, a: string, b: string) => {
  if (!isString(s)) return "";
  if (!isString(a)) return s;
  if (!isString(b)) b = "";
  return s.replaceAll(a, b);
});

handlebars.registerHelper("replace", (s: string, a: string, b: string) => {
  if (!isString(s)) return "";
  if (!isString(a)) return s;
  if (!isString(b)) b = "";
  return s.replace(a, b);
});

handlebars.registerHelper("replace", (s: string, a: string, b: string) => {
  if (!isString(s)) return "";
  if (!isString(a)) return s;
  if (!isString(b)) b = "";
  return s.replace(a, b);
});

handlebars.registerHelper("rbase64", (src: string) => {
  let gitRoot = findUpSync(".git");
  if (!gitRoot) {
    gitRoot = findUpSync(".git", { type: "directory" } as Options);
    if (!gitRoot) {
      return "Outside is only applicable in git projects.";
    }
  }
  let p = path.join(path.dirname(gitRoot), src);
  const fileDataBase64 = fs.readFileSync(p, "base64");
  const mime = lookup(p);
  return `data:${mime};base64,${fileDataBase64}`;
});

handlebars.registerHelper(
  "rfile",
  (src: string, lines?: string, ellipsis?: string) => {
    if (!src) {
      throw Error("file needs a path to a file");
    }
    let gitRoot = findUpSync(".git");
    if (!gitRoot) {
      gitRoot = findUpSync(".git", { type: "directory" } as Options);
      if (!gitRoot) {
        return "Outside is only applicable in git projects.";
      }
    }
    let p = path.join(path.dirname(gitRoot), src);
    const content = fs.readFileSync(p, "utf8");
    return extractLines(content, lines, ellipsis);
  }
);

export { handlebars };
