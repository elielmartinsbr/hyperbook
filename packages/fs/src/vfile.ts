import path from "path";
import matter from "gray-matter";
import fs from "fs/promises";
import fsD from "fs";
import { HyperbookFrontmatter } from "@hyperbook/types";
import yaml from "yaml";
import { handlebars, registerHelpers } from "./handlebars";
import { lookup } from "mime-types";

export type VFile = {
  root: string;
  folder: "archives" | "glossary" | "book" | "public" | "snippets";
  path: {
    directory: string;
    relative: string;
    absolute: string;
    href: string | null;
  };
  name: string;
  extension: string;
};

export type VDirectory = {
  root: string;
  folder: "archives" | "glossary" | "book" | "public" | "snippets";
  name: string;
  path: {
    relative: string;
    absolute: string;
  };
  files: VFile[];
  index?: VFile;
  directories: VDirectory[];
};

const folders = ["archives", "glossary", "book", "public", "snippets"] as const;

export function getHref(vfile: {
  path: { relative: string; directory: string };
  folder: VFile["folder"];
  name: VFile["name"];
  extension: VFile["extension"];
}): VFile["path"]["href"] {
  switch (vfile.folder) {
    case "book": {
      let href = "/";
      if (vfile.path.directory) {
        href = path.join(href, vfile.path.directory);
      }
      if (vfile.name !== "index") {
        href = path.join(href, vfile.name);
      }
      return href.trim();
    }
    case "snippets": {
      return null;
    }
    case "glossary": {
      let href = "/glossary";
      if (vfile.path.directory) {
        href = path.join(href, vfile.path.directory);
      }
      if (vfile.name !== "index") {
        href = path.join(href, vfile.name);
      }
      return href.trim();
    }
    case "public": {
      return "/" + vfile.path.relative;
    }
    case "archives": {
      return "/archives/" + vfile.path.directory + ".zip";
    }
  }
}

export async function getDirectory(
  root: string,
  folder: VFile["folder"],
  extensions?: string[]
): Promise<VDirectory> {
  async function getTree(directory: VDirectory): Promise<VDirectory> {
    const files = await fs
      .readdir(path.join(directory.path.absolute))
      .catch(() => []);
    for (const file of files) {
      const stat = await fs.stat(path.join(directory.path.absolute, file));
      if (stat.isDirectory()) {
        const { name } = path.parse(file);
        const d: VDirectory = {
          name,
          root,
          folder,
          directories: [],
          files: [],
          path: {
            absolute: path.join(directory.path.absolute, file),
            relative: path.join(directory.path.relative, file),
          },
        };
        directory.directories.push(await getTree(d));
      } else {
        let { ext, name } = path.parse(file);
        if (ext == ".hbs" && file.endsWith(".md.hbs")) {
          ext = ".md.hbs";
          name = name.slice(0, name.length - 3);
        }
        if (extensions && !extensions.includes(ext)) {
          continue;
        }
        let vfile: VFile = {
          folder: folder as VFile["folder"],
          path: {
            directory: directory.path.relative,
            absolute: path.join(directory.path.absolute, file),
            relative: path.join(directory.path.relative, file),
            href: "",
          },
          name,
          root,
          extension: ext,
        };
        vfile.path.href = getHref(vfile);
        if (name === "index") {
          directory.index = vfile;
        } else {
          directory.files.push(vfile);
        }
      }
    }
    return directory;
  }

  return getTree({
    root,
    path: {
      relative: "",
      absolute: path.join(root, folder),
    },
    name: "",
    folder: folder,
    directories: [],
    files: [],
  });
}

export const flatDirectory = async ({
  directories,
  files,
  index,
}: VDirectory): Promise<VFile[]> => {
  const allFiles: VFile[] = [...files];
  if (index) {
    allFiles.push(index);
  }

  for (const directory of directories) {
    const dirFiles = await flatDirectory(directory);
    allFiles.push(...dirFiles);
  }

  return allFiles;
};

export async function listForFolder(
  root: string,
  folder: VFile["folder"],
  extension?: string[]
): Promise<VFile[]> {
  const directory = await getDirectory(root, folder, extension);
  return flatDirectory(directory);
}

export const list = async (
  root: string,
  extension?: string[]
): Promise<VFile[]> => {
  return Promise.all(
    folders.flatMap((folder) => listForFolder(root, folder, extension))
  ).then((f) => f.flat());
};

export const get = async (
  root: string,
  folder: VFile["folder"] | null,
  href: string
): Promise<VFile | undefined> => {
  if (folder === null) {
    const files = await list(root);
    return files.find((f) => f.path.href === href);
  }
  const files = await listForFolder(root, folder);
  return files.find((f) => f.path.href === href);
};

export const getByAbsolutePath = async (
  root: string,
  folder: VFile["folder"] | null,
  absolute: string
) => {
  if (folder === null) {
    const files = await list(root);
    return files.find((f) => f.path.absolute === absolute);
  }
  const files = await listForFolder(root, folder);
  return files.find((f) => f.path.absolute === absolute);
};

export const extractLines = (
  content: string,
  lines?: string,
  ellipsis?: string
): string => {
  if (!lines || typeof lines !== "string") {
    return content;
  }

  let extractedLines: string[] = [];
  if (!lines.startsWith("reg:")) {
    const lineNumbers: number[] = [];
    for (const line of lines.split(",")) {
      if (line.includes("-")) {
        const [start, end] = line.split("-");
        for (let i = Number(start); i <= Number(end); i++) {
          lineNumbers.push(i);
        }
      } else {
        lineNumbers.push(Number(line));
      }
    }

    const contentLines = content.split("\n");
    let lastLineNumber = 0;
    for (const lineNumber of lineNumbers) {
      if (lineNumber - lastLineNumber > 1 && ellipsis) {
        extractedLines.push(ellipsis);
      }
      extractedLines.push(contentLines[lineNumber - 1]);
      lastLineNumber = lineNumber;
    }
    if (lastLineNumber < contentLines.length - 1 && ellipsis) {
      extractedLines.push(ellipsis);
    }
  } else {
    const contentLines = content.split("\n");
    const reg = new RegExp(lines.slice(4));
    extractedLines = contentLines.filter((c) => !reg.test(c));
  }

  return extractedLines.join("\n");
};

export const getMarkdown = async (
  file: VFile
): Promise<{ content: string; data: HyperbookFrontmatter }> => {
  if (file.folder !== "book" && file.folder !== "glossary") {
    throw Error(
      `Unsupported file location. Only files from book and glossary are supported for reading their markdown content.`
    );
  }

  registerHelpers(handlebars, { file });

  let markdown = "";
  if (file.extension === ".yml") {
    const j = await fs
      .readFile(file.path.absolute)
      .then((f) => yaml.parse(f.toString()));
    if (!j.template) {
      console.log(
        `Yaml files need a template. You need to define a template property in your yaml file: ${file.path.absolute}`
      );
    } else {
      const templateSource = await fs.readFile(
        path.join(file.root, "templates", j.template + ".md.hbs")
      );
      const template = handlebars.compile(templateSource.toString());
      markdown = template(j);
    }
  } else if (file.extension === ".json") {
    const j = await fs
      .readFile(file.path.absolute)
      .then((f) => JSON.parse(f.toString()))
      .catch(() => ({}));
    if (!j.template) {
      console.log(
        `JSON files need a template. You need to define a template property in your json file: ${file.path.absolute}`
      );
    } else {
      const templateSource = await fs.readFile(
        path.join(file.root, "templates", j.template + ".md.hbs")
      );
      const template = handlebars.compile(templateSource.toString());
      markdown = template(j);
    }
  } else if (file.extension === ".md") {
    markdown = await fs.readFile(file.path.absolute).then((f) => f.toString());
  } else if (file.extension === ".md.hbs") {
    const templateString = await fs
      .readFile(file.path.absolute)
      .then((f) => f.toString());
    const template = handlebars.compile(templateString);
    markdown = template();
  } else {
    console.log(
      `Unsupported file extension ${file.extension}. Only .md, .yml, .json and .md.hbs files are supported.`
    );
  }
  let { content, data } = matter(markdown);

  // apply Snippets
  const reg =
    /((:+)snippet{#(?<snippet>[a-zA-Z0-9-]+)(?<vars>( +(?<var>[a-zA-Z]+)=(?<value>\d+|false|true|".*"))*) *})/g;
  const varReg = /(?<var>[a-zA-Z]+)=(?<value>\d+|false|true|".*")/;
  const snippets = [...content.matchAll(reg)];

  for (const snippet of snippets) {
    const dots = snippet[2];
    const snippetId = snippet[3];
    const snippetFile = await fs.readFile(
      path.join(file.root, "snippets", snippetId + ".md.hbs"),
      { encoding: "utf8" }
    );
    const template = handlebars.compile(snippetFile);
    const vars = {};
    for (const m of snippet[4].split(" ")) {
      const r = varReg.exec(m);
      if (r) {
        if (r[2].startsWith('"') && r[2].endsWith('"')) {
          vars[r[1]] = r[2].slice(1, -1);
        } else if (r[2] === "false") {
          vars[r[1]] = false;
        } else if (r[2] === "true") {
          vars[r[1]] = true;
        } else {
          vars[r[1]] = Number(r[2]);
        }
      }
    }
    if (dots.length > 2) {
      const r = `${snippet[0]}([\\s\\S]*?)${dots}`;
      const m = content.match(new RegExp(r, "m"));
      if (m) {
        vars["content"] = m[1];
        content = content.replace(m[0], template(vars));
      }
    } else {
      content = content.replace(snippet[0], template(vars));
    }
  }

  return { content, data: { ...data } as HyperbookFrontmatter };
};
