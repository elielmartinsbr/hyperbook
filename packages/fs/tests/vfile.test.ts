import path from "path";
import { describe, it, expect } from "vitest";
import * as vfile from "../src/vfile";

describe("list", () => {
  it("should list all", async () => {
    let websiteEn = path.join(__dirname, "..", "..", "..", "website", "en");
    expect(await vfile.list(websiteEn)).toMatchSnapshot();
  });

  it("should list for folder book", async () => {
    let websiteEn = path.join(__dirname, "..", "..", "..", "website", "en");
    expect(await vfile.listForFolder(websiteEn, "book")).toMatchSnapshot();
  });

  it("should include main index", async () => {
    let websiteEn = path.join(__dirname, "..", "..", "..", "website", "en");
    const files = await vfile.listForFolder(websiteEn, "book");
    const file = files.find((f) => f.path.href === "/");
    expect(file).toBeDefined();
  });
});

describe("getMarkdown", () => {
  it("should get markdown from template", async () => {
    let websiteEn = path.join(__dirname, "..", "..", "..", "website", "en");
    let files = await vfile.list(websiteEn);
    let templateFile = files.find(
      (f) => f.name === "template-demo-yaml" && f.folder === "book"
    );
    if (!templateFile) {
      throw Error("Template not found");
    }

    expect(await vfile.getMarkdown(templateFile)).toMatchSnapshot();
  });
});

describe("getDirectory", () => {
  it("should get directory", async () => {
    let websiteEn = path.join(__dirname, "..", "..", "..", "website", "en");
    let directory = await vfile.getDirectory(websiteEn, "book");

    expect(directory).toMatchSnapshot();
  });

  it("should include main index", async () => {
    let websiteEn = path.join(__dirname, "..", "..", "..", "website", "en");
    let directory = await vfile.getDirectory(websiteEn, "book");
    expect(directory.index).toBeDefined();
  });
});

describe("extractLines", () => {
  const text =
    "Line 1\nLine 2\nLine 3\nLine 4\nLine 5\nLine 6\nLine 7\nLine 8\nLine 9\nLine 10";
  it("should extract lines by range", () => {
    const extractedLines = vfile.extractLines(text, "2-4", "...");
    expect(extractedLines).toMatchSnapshot();
  });
  it("should extract lines with gaps", () => {
    const extractedLines = vfile.extractLines(text, "1,4-6,8", "...");
    expect(extractedLines).toMatchSnapshot();
  });
  it("should extract lines with no ellipsis", () => {
    const extractedLines = vfile.extractLines(text, "1,4-6,8");
    expect(extractedLines).toMatchSnapshot();
  });
  it("should extract lines with custom ellipsis", () => {
    const extractedLines = vfile.extractLines(text, "1,4-6,8", "// ...");
    expect(extractedLines).toMatchSnapshot();
  });
  it("should extract lines and not add ellipsis on first and last line", () => {
    const extractedLines = vfile.extractLines(text, "1,4-6,10", "// ...");
    expect(extractedLines).toMatchSnapshot();
  });
  it("should return full content", () => {
    const extractedLines = vfile.extractLines(text);
    expect(extractedLines).toMatchSnapshot();
  });
});
