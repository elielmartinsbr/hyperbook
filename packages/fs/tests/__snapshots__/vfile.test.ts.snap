// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extractLines > should extract lines and not add ellipsis on first and last line 1`] = `
"Line 1
// ...
Line 4
Line 5
Line 6
// ...
Line 10"
`;

exports[`extractLines > should extract lines by range 1`] = `
"...
Line 2
Line 3
Line 4
..."
`;

exports[`extractLines > should extract lines with custom ellipsis 1`] = `
"Line 1
// ...
Line 4
Line 5
Line 6
// ...
Line 8
// ..."
`;

exports[`extractLines > should extract lines with gaps 1`] = `
"Line 1
...
Line 4
Line 5
Line 6
...
Line 8
..."
`;

exports[`extractLines > should extract lines with no ellipsis 1`] = `
"Line 1
Line 4
Line 5
Line 6
Line 8"
`;

exports[`extractLines > should return full content 1`] = `
"Line 1
Line 2
Line 3
Line 4
Line 5
Line 6
Line 7
Line 8
Line 9
Line 10"
`;

exports[`getMarkdown > should get markdown for single use template 1`] = `
{
  "content": "# Single Use Template

TEST

Hi!
",
  "data": {},
}
`;

exports[`getMarkdown > should get markdown from template 1`] = `
{
  "content": "
Example use of a template.
",
  "data": {
    "name": "Use Template1 with YAML",
  },
}
`;

exports[`getMarkdown > should get markdown from template for index 1`] = `
{
  "content": "
Example use of a template.
",
  "data": {
    "name": "Use Template1 with YAML",
  },
}
`;

exports[`list > should list all 1`] = `
[
  {
    "extension": ".md",
    "folder": "glossary",
    "name": "oop",
    "path": {
      "absolute": "single-hyperbook/glossary/oop.md",
      "directory": "",
      "href": "/glossary/oop",
      "relative": "oop.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": "",
    "folder": "book",
    "name": ".gitkeep",
    "path": {
      "absolute": "single-hyperbook/book/.gitkeep",
      "directory": "",
      "href": "/.gitkeep",
      "relative": ".gitkeep",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "name": "paradigms",
    "path": {
      "absolute": "single-hyperbook/book/paradigms.md",
      "directory": "",
      "href": "/paradigms",
      "relative": "paradigms.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/index.md",
      "directory": "",
      "href": "/",
      "relative": "index.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".json",
    "folder": "book",
    "name": "empty",
    "path": {
      "absolute": "single-hyperbook/book/subsection/empty.json",
      "directory": "subsection",
      "href": "/subsection/empty",
      "relative": "subsection/empty.json",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".json",
    "folder": "book",
    "name": "invalid",
    "path": {
      "absolute": "single-hyperbook/book/subsection/invalid.json",
      "directory": "subsection",
      "href": "/subsection/invalid",
      "relative": "subsection/invalid.json",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".yml",
    "folder": "book",
    "name": "use-template1",
    "path": {
      "absolute": "single-hyperbook/book/subsection/use-template1.yml",
      "directory": "subsection",
      "href": "/subsection/use-template1",
      "relative": "subsection/use-template1.yml",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/subsection/index.md",
      "directory": "subsection",
      "href": "/subsection",
      "relative": "subsection/index.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "name": "test",
    "path": {
      "absolute": "single-hyperbook/book/subsection2/test.md",
      "directory": "subsection2",
      "href": "/subsection2/test",
      "relative": "subsection2/test.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".yml",
    "folder": "book",
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/subsection2/index.yml",
      "directory": "subsection2",
      "href": "/subsection2",
      "relative": "subsection2/index.yml",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md.hbs",
    "folder": "book",
    "name": "single-template",
    "path": {
      "absolute": "single-hyperbook/book/subsection3/single-template.md.hbs",
      "directory": "subsection3",
      "href": "/subsection3/single-template",
      "relative": "subsection3/single-template.md.hbs",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".yml",
    "folder": "book",
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/subsection3/index.yml",
      "directory": "subsection3",
      "href": "/subsection3",
      "relative": "subsection3/index.yml",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "name": "test",
    "path": {
      "absolute": "single-hyperbook/book/subsection3/subsection31/test.md",
      "directory": "subsection3/subsection31",
      "href": "/subsection3/subsection31/test",
      "relative": "subsection3/subsection31/test.md",
    },
    "root": "single-hyperbook",
  },
]
`;

exports[`list > should list for folder book 1`] = `
[
  {
    "extension": "",
    "folder": "book",
    "name": ".gitkeep",
    "path": {
      "absolute": "single-hyperbook/book/.gitkeep",
      "directory": "",
      "href": "/.gitkeep",
      "relative": ".gitkeep",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "name": "paradigms",
    "path": {
      "absolute": "single-hyperbook/book/paradigms.md",
      "directory": "",
      "href": "/paradigms",
      "relative": "paradigms.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/index.md",
      "directory": "",
      "href": "/",
      "relative": "index.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".json",
    "folder": "book",
    "name": "empty",
    "path": {
      "absolute": "single-hyperbook/book/subsection/empty.json",
      "directory": "subsection",
      "href": "/subsection/empty",
      "relative": "subsection/empty.json",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".json",
    "folder": "book",
    "name": "invalid",
    "path": {
      "absolute": "single-hyperbook/book/subsection/invalid.json",
      "directory": "subsection",
      "href": "/subsection/invalid",
      "relative": "subsection/invalid.json",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".yml",
    "folder": "book",
    "name": "use-template1",
    "path": {
      "absolute": "single-hyperbook/book/subsection/use-template1.yml",
      "directory": "subsection",
      "href": "/subsection/use-template1",
      "relative": "subsection/use-template1.yml",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/subsection/index.md",
      "directory": "subsection",
      "href": "/subsection",
      "relative": "subsection/index.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "name": "test",
    "path": {
      "absolute": "single-hyperbook/book/subsection2/test.md",
      "directory": "subsection2",
      "href": "/subsection2/test",
      "relative": "subsection2/test.md",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".yml",
    "folder": "book",
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/subsection2/index.yml",
      "directory": "subsection2",
      "href": "/subsection2",
      "relative": "subsection2/index.yml",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md.hbs",
    "folder": "book",
    "name": "single-template",
    "path": {
      "absolute": "single-hyperbook/book/subsection3/single-template.md.hbs",
      "directory": "subsection3",
      "href": "/subsection3/single-template",
      "relative": "subsection3/single-template.md.hbs",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".yml",
    "folder": "book",
    "name": "index",
    "path": {
      "absolute": "single-hyperbook/book/subsection3/index.yml",
      "directory": "subsection3",
      "href": "/subsection3",
      "relative": "subsection3/index.yml",
    },
    "root": "single-hyperbook",
  },
  {
    "extension": ".md",
    "folder": "book",
    "name": "test",
    "path": {
      "absolute": "single-hyperbook/book/subsection3/subsection31/test.md",
      "directory": "subsection3/subsection31",
      "href": "/subsection3/subsection31/test",
      "relative": "subsection3/subsection31/test.md",
    },
    "root": "single-hyperbook",
  },
]
`;
